File: backend/dotnet/clean-architecture/minimal/src/Presentation/App.Template.WebAPI/Controllers/HealthController.cs
using AppTemplate.Application.Interfaces;
using Microsoft.AspNetCore.Mvc;

namespace AppTemplate.WebAPI.Controllers;

/// <summary>
/// Health check endpoints for monitoring and orchestration (Kubernetes, Docker, etc.)
/// </summary>
[ApiController]
[Route("[controller]")]
public class HealthController : ControllerBase
{
    private readonly IApplicationDbContext _context;

    public HealthController(IApplicationDbContext context)
    {
        _context = context;
    }

    /// <summary>
    /// Basic health check - returns application status
    /// </summary>
    [HttpGet("/health")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    public IActionResult Health()
    {
        return Ok(new
        {
            status = "healthy",
            timestamp = DateTime.UtcNow,
            application = "AppTemplate API",
            version = "1.0.0"
        });
    }

    /// <summary>
    /// Readiness check - verifies the application can accept traffic (database connected)
    /// Returns 503 if dependencies are not ready
    /// </summary>
    [HttpGet("/health/ready")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status503ServiceUnavailable)]
    public async Task<IActionResult> Ready()
    {
        try
        {
            // Check database connectivity
            var canConnect = await _context.Database.CanConnectAsync();

            if (canConnect)
            {
                return Ok(new
                {
                    status = "ready",
                    timestamp = DateTime.UtcNow,
                    database = "connected"
                });
            }
        }
        catch
        {
            // Database connection failed
        }

        return StatusCode(StatusCodes.Status503ServiceUnavailable, new
        {
            status = "not ready",
            timestamp = DateTime.UtcNow,
            database = "disconnected"
        });
    }

    /// <summary>
    /// Liveness check - verifies the application is alive
    /// Always returns 200 if the application is responding
    /// </summary>
    [HttpGet("/health/live")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    public IActionResult Live()
    {
        return Ok(new
        {
            status = "alive",
            timestamp = DateTime.UtcNow
        });
    }
}

File: backend/nestjs/feature-architecture/full/src/features/health/health.controller.ts
import { Controller, Get, HttpStatus, Res } from '@nestjs/common';
import { ApiOperation, ApiResponse, ApiTags } from '@nestjs/swagger';
import { DataSource } from 'typeorm';
import { Response } from 'express';
import { SkipTransform } from '../../common/decorators/response-message.decorator';

@Controller('health')
@ApiTags('Health')
export class HealthController {
  constructor(private readonly dataSource: DataSource) {}

  @Get()
  @SkipTransform()
  @ApiOperation({ summary: 'Basic health check' })
  @ApiResponse({ status: 200, description: 'Application is healthy' })
  health() {
    return {
      status: 'healthy',
      timestamp: new Date().toISOString(),
    };
  }

  @Get('ready')
  @SkipTransform()
  @ApiOperation({ summary: 'Readiness check' })
  @ApiResponse({ status: 200, description: 'Application is ready' })
  @ApiResponse({ status: 503, description: 'Application is not ready' })
  async ready(@Res() res: Response) {
    try {
      await this.dataSource.query('SELECT 1');
      return res.status(HttpStatus.OK).json({
        status: 'ready',
        timestamp: new Date().toISOString(),
        database: 'connected',
      });
    } catch {
      return res.status(HttpStatus.SERVICE_UNAVAILABLE).json({
        status: 'not ready',
        timestamp: new Date().toISOString(),
        database: 'disconnected',
      });
    }
  }

  @Get('live')
  @SkipTransform()
  @ApiOperation({ summary: 'Liveness check' })
  @ApiResponse({ status: 200, description: 'Application is alive' })
  live() {
    return {
      status: 'alive',
      timestamp: new Date().toISOString(),
    };
  }
}
